fastestimator/dataset/batch_dataset.py

```diff
@@ -0,0 +1,132 @@
+# Copyright 2019 The FastEstimator Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+import math
+import random
+from typing import Any, Dict, Iterable, List, Optional, Sequence, Union
+
+import numpy as np
+
+from fastestimator.dataset.dataset import DatasetSummary, FEDataset
+from fastestimator.util import to_list
+
+
+class BatchDataset(FEDataset):
+    """BatchDataset can create list of batch data from single dataset or multiple datasets.
+
+    Args:
+        datasets: FEDataset instance or list of FEDataset instances.
+        num_samples: Number of samples to draw from dataset(s).
+        probability: Probability to draw from each dataset. Defaults to None.
+
+        Example:
+        Given two datasets if the batch size is 8 and we want to randomly draw 5 from first dataset and remaining 3 from
+         the second: num_samples = [5, 3].
+
+        Given two datasets if the batch size is 8 and we want to draw with 0.7 probability from first dataset and 0.3
+        from the second: num_samples = 8, probability = [0.7, 0.3].
+    """
+    def __init__(self,
+                 datasets: Union[FEDataset, Iterable[FEDataset]],
+                 num_samples: Union[int, Iterable[int]],
+                 probability: Optional[Iterable[float]] = None):
+        self.datasets = to_list(datasets)
+        self.num_samples = to_list(num_samples)
+        self.probability = to_list(probability)
+        self.same_feature = False
+        self._check_input()
+        self.index_maps = [list(range(max((len(dataset) for dataset in self.datasets)))) for _ in self.datasets]
```

Comment:
technically you might want to compute the max first to reduce looping, but realistically this won't matter